### PR TITLE
Epmcuxda 6912: aggregations + taking in account user role

### PR DESCRIPTION
**Pull request by SerafimovichEugene**
Aggregations now depend on user role. 
PS. Getting aggregations is not really efficient ( O(n*m*(3*k)) ) as it is getting now them by mapping and reducing observations. For a big chunk of observations, especially for admin (all observations)
it would be probably too slow.

### Jira ticket related to this PR

- [EPMCUXDA-6912](https://jira.epam.com/jira/browse/EPMCUXDA-6912)

#### Type of change

`What types of changes does your code introduce? Put an 'x' in all the boxes that apply:`

- [x] Adds new features.
- [ ] Changes existing functionality.
- [ ] Deprecates features.
- [ ] Removes features.
- [ ] Fixes any bug.
- [ ] Other: security, build process, infrastructure, tests, docs, etc.

### CODE REVIEW

File: src/controllers/observation-controller.ts [hunk 1]
```diff
@@ -63,29 +72,18 @@ export default class ObservationController extends AbstractController {
       const langOrigin = LocaleOrigin[lang] ? LocaleOrigin[lang] : 'desc_eng';
 
       const paramsSearch = parsePageParams(req.query);
-      const paramsAggregation = parseWhereParams(req.query, req.user);
+      const paramsAggregation = parseWhereParams(req.user, req.query);
       const observations = await this.observations.findAndCount(Object.assign(paramsSearch, paramsAggregation));
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: please, use here de-structuring to improve code readability

File: src/controllers/observation-controller.ts [hunk 2]
```diff
@@ -104,10 +102,42 @@ export default class ObservationController extends AbstractController {
     }
   };
 
-  private getAggregations = async (_req: Request, res: Response, next: NextFunction) => {
+  private getAggregations = async (req: Request, res: Response, next: NextFunction) => {
     try {
-      const aggregations = await getAggregations(this.observations);
-      res.json({ ...aggregations });
+      const { lang = 'eng' }: { lang: string } = req.query;
+      const langOrigin = LocaleOrigin[lang] ? LocaleOrigin[lang] : 'desc_eng';
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: Would be better to simplify with logical OR

File: src/controllers/observation-controller.ts [hunk 3]
```diff
@@ -104,10 +102,42 @@ export default class ObservationController extends AbstractController {
...
+
+      const paramsAggregation = parseWhereParams(req.user, req.query);
+      const observations = await this.observations.find({ ...paramsAggregation });
+      const aggregatinMap: AggregationsMap = {};
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: Redundant annotation, please merge with next line

File: src/controllers/observation-controller.ts [hunk 4]
```diff
@@ -63,29 +72,18 @@ export default class ObservationController extends AbstractController {
...
-                  subfield === langOrigin ? ['desc', subValue] : ([subfield, subValue] as any),
-                )
-                .reduce((acc, [subfield, subValue]) => Object.assign(acc, { [subfield]: subValue }), {});
-              return [ObservationKey, value];
-            }
-            return [ObservationKey, ObservationValue];
-          })
+          .map(entrie => mapLocale(entrie, langOrigin))
           .reduce(
             (acc, [ObservationField, ObservationValue]) => Object.assign(acc, { [ObservationField]: ObservationValue }),
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: Could be just `[key, value]`, as currently too verbose and obvious

File: src/controllers/observation-controller.ts [hunk 5]
```diff
@@ -63,29 +72,18 @@ export default class ObservationController extends AbstractController {
...
       const observations = await this.observations.findAndCount(Object.assign(paramsSearch, paramsAggregation));
 
       const content = observations[0].map(obs => {
         // sanitaze user's data
-        const finder = Object.assign({}, obs.finder.sanitizeUser(), { id: obs.finder.id });
+        const finder = Object.assign({}, User.sanitizeUser(obs.finder));
         // transform 'observation'
         const observation = Object.entries(obs)
           // clear 'filter' field
           .filter(([ObservationField]) => ObservationField !== 'ring')
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: just `field`

File: src/controllers/observation-controller.ts [hunk 6]
```diff
@@ -63,29 +72,18 @@ export default class ObservationController extends AbstractController {
       const langOrigin = LocaleOrigin[lang] ? LocaleOrigin[lang] : 'desc_eng';
 
       const paramsSearch = parsePageParams(req.query);
-      const paramsAggregation = parseWhereParams(req.query, req.user);
+      const paramsAggregation = parseWhereParams(req.user, req.query);
       const observations = await this.observations.findAndCount(Object.assign(paramsSearch, paramsAggregation));
 
       const content = observations[0].map(obs => {
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: try to use or on letter like `o` or full key `observation` without inter-immediate variants, with one known and widely used shorhand for `reduce`: `accumulator` = > `acc | accum | others`

File: src/services/observation-service.ts [hunk 7]
```diff
@@ -48,33 +48,31 @@ export const filterFieldByLocale = (key: Locale, lang: Locale): boolean => {
...
-      .getRawMany();
-    return reduceWithCount(res, 'finder');
-  },
-];
-
-export const parseWhereParams = (query: ObservationAggregations, user: User): FindOneOptions<Observation> => {
+
+export const parseWhereParams = (user: User, query: ObservationAggregations): FindOneOptions<Observation> => {
   const params = Object.entries(query)
     .filter(entrie => !aggregationSearch.includes(entrie[0])) // filter search aggregations, only ObservationAggregations could be applied
```
Comment (mishelen) [inline src/services/observation-service.ts]: for both `map` and `filter` is preferable to use destructuring to increase readability

File: src/controllers/observation-controller.ts (hunk 3 above)
Comment (mishelen) [reply src/controllers/observation-controller.ts]: Don't pay attention that `outdated`, it's because my commit. Still actual

Comment (SerafimovichEugene) [reply src/controllers/observation-controller.ts]: couldn't remove it due to ts ling error.

Comment (mishelen) [reply src/controllers/observation-controller.ts]: Please take a look: 49c42d6 -- no any error
@SerafimovichEugene

File: src/controllers/observation-controller.ts (hunk 1 above)
Comment (mishelen) [reply src/controllers/observation-controller.ts]: I'd asked for this change: c3e4aae 
@SerafimovichEugene

File: src/controllers/observation-controller.ts [hunk 8]
```diff
@@ -54,60 +62,81 @@ export default class ObservationController extends AbstractController {
...
-              return [ObservationKey, value];
-            }
-            return [ObservationKey, ObservationValue];
-          })
-          .reduce(
-            (acc, [ObservationField, ObservationValue]) => Object.assign(acc, { [ObservationField]: ObservationValue }),
-            {},
-          );
-        return Object.assign({}, observation, { finder }, { ring: obs.ring.id });
+          .map(entry => mapLocale(entry, langOrigin))
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: `mapLocale` should be unaware about exact field to choose, I mean this method should decide this internally and never inside calling it methods

Comment (mishelen) [reply src/controllers/observation-controller.ts]: This is kind of broken encapsulation which in current case will cause code duplication in future

File: src/controllers/observation-controller.ts (hunk 1 above)
Comment (SerafimovichEugene) [reply src/controllers/observation-controller.ts]: now I see))

File: src/controllers/observation-controller.ts (hunk 8 above)
Comment (SerafimovichEugene) [reply src/controllers/observation-controller.ts]: yeah, got it. Now 'query' is passed to mapLocale as the second argument.

**Comment by SerafimovichEugene** [comment]
> Hey, Eugene. I've fixed some issues to not waste your time on communication, but, please, go through my prev. comments where I've pointed on commits and resolve them.
> 
> While I was looking through code I found 2 places to improve:
> 
> * **first we are using as `lang` values like `desc_rus`... It should be `rus`.** I had warned you on that before. Let me share my some considerations: you expose externally you internal values, but this don't afraid me as fact that anytime when we'll have other fields to choose like `title_rus`, `code_rus` ...  **you'll know right away how wrong this api parameter was**
> * second, observations service shouldn't store language related typings, please move all of them to `entities/common-interfaces`
> * plus 1 new  related to `lang` note in comments
> 
> Still actual note about params naming, mentioned by me in previous request,
> 
> > Also I couldn't reflect in comments, but there is some big cognitive load with all this searchParams, whereParams, Aggeragations and other related variables, functions, types and so on. I am mostly complaining that they seems as unrelated to each other.
> 
> seems that we need discuss this offline or other way

I didn't understand what did you mean in the first paragraph. I use  `lang` values like `eng`, `byn` or `rus`. Then I just try to match `lang` param with the *column name* (desc_eng, etc.) and if it fails I just use `eng` by default.

File: src/controllers/observation-controller.ts [hunk 9]
```diff
@@ -54,60 +61,76 @@ export default class ObservationController extends AbstractController {
...
+            return item.value === observation[column];
+          });
+          if (desired) {
+            desired.count += 1;
+          } else {
+            const f = pipe(
+              (arg: [string, any]) => mapLocale(arg, query),
+              (arg: [string, any]) => sanitizeUser(arg),
+            );
+            const [, obsValue] = f([column, observation[column]]);
```
Comment (mishelen) [inline src/controllers/observation-controller.ts]: `obsValue` is too verbose and obvious just `value`

**Pull request merged**